### PR TITLE
Add the Morello functions to interact with userspace memory

### PR DESCRIPTION
**Pull request by zxombie**
This adds the cap variants of the copyin/copyout and fue/su functions

### CODE REVIEW

File: sys/arm64/arm64/copyinout.S [hunk 1]
```diff
@@ -224,3 +256,172 @@ ending:
...
+	sub	x2, x2, #1
+	add	x5, x5, #1
+	cbnz	x2, captag_end_bytes
+
+captag_done:
+	EXIT_USER_ACCESS(w6)
+	SET_FAULT_HANDLER(xzr, x7) /* Clear the handler */
+
+	mov	x0, xzr		/* return 0 */
+	RET
```
Comment (jrtc27) [inline sys/arm64/arm64/copyinout.S:382]:
```suggestion
	ret
```

File: sys/arm64/arm64/copyinout.S [hunk 2]
```diff
@@ -91,23 +105,40 @@ END(copyin)
...
 	ENTER_USER_ACCESS(w6, x7)
 
 	ldr	x7, =VM_MAXUSER_ADDRESS
-1:	cmp	x0, x7
+1:
+#if __has_feature(capabilities)
+	gcvalue	x9, c0
+	cmp	x9, x7
+	b.cs	copyio_fault
+	ldrb	w4, [c0]	/* Load from uaddr */
```
Comment (arichardson) [inline sys/arm64/arm64/copyinout.S:131]: Do we want to also use ldtrb here? We could do a bx at the start+end for hybrid kernels?

Comment (arichardson) [reply sys/arm64/arm64/copyinout.S:131]: (and a `.arch morello+c64` directive)

File: sys/arm64/arm64/support.S [hunk 3]
```diff
@@ -38,32 +38,53 @@ __FBSDID("$FreeBSD$");
 
 #include "assym.inc"
 
+#if __has_feature(capabilities)
```
Comment (arichardson) [inline sys/arm64/arm64/support.S:41]:
```suggestion
#if __has_feature(capabilities) && !defined __CHERI_PURE_CAPABILITY__
```

File: sys/arm64/arm64/support.S [hunk 4]
```diff
@@ -38,32 +38,53 @@ __FBSDID("$FreeBSD$");
...
  */
 ENTRY(fsu_fault)
+	SWITCH_MODE
+EENTRY(fsu_noswitch_fault)
 	SET_FAULT_HANDLER(xzr, x1)	/* Reset the handler function */
 	EXIT_USER_ACCESS_CHECK(w0, x1)
 fsu_fault_nopcb:
 	mov	x0, #-1
 	ret
+EEND(fsu_a64_fault)
```
Comment (arichardson) [inline sys/arm64/arm64/support.S:69]:
```suggestion
EEND(fsu_noswitch_fault)
```

File: sys/arm64/arm64/support.S (hunk 3 above)
Comment (zxombie) [reply sys/arm64/arm64/support.S:41]: That looks correct, however there are other issues with the code so I's unlikely to work in a purecap kernel as is.

**Comment by zxombie** [comment]
The `SWITCH_MODE` will enter C64 mode where the encoding for to load via an integer register is uses as a load via a capability register. This means these memory accesses, while they use an `x` register use the `c` register.

**Comment by jrtc27** [comment]
> The `SWITCH_MODE` will enter C64 mode where the encoding for to load via an integer register is uses as a load via a capability register. This means these memory accesses, while they use an `x` register use the `c` register.

Yeah thinking about its more I came to realise that might be it. Do I really need to say what my thoughts are around _that_?

**Comment by zxombie** [comment]
> Can this not all be made to work by temporarily putting the assembler in C64 mode? Then the symbol will have its LSB set so branches to it will automatically work.

I'm not sure if there is interworking support in the compiler between A64 and C64. Most calls will use a `BL` instruction which doesn't store the bottom 2 bits of the address so can't provide the interworking needed.

**Comment by jrtc27** [comment]
> > Can this not all be made to work by temporarily putting the assembler in C64 mode? Then the symbol will have its LSB set so branches to it will automatically work.
> 
> I'm not sure if there is interworking support in the compiler between A64 and C64. Most calls will use a `BL` instruction which doesn't store the bottom 2 bits of the address so can't provide the interworking needed.

LLD looks like it has code for `R_{AARCH64,MORELLO}_{CALL,JUMP}26` to go via thunks if calling a symbol with the "wrong" LSB.

**Comment by arichardson** [comment]
> The `SWITCH_MODE` will enter C64 mode where the encoding for to load via an integer register is uses as a load via a capability register. This means these memory accesses, while they use an `x` register use the `c` register.

Doesn't `.arch morello+c64` work? Looking at the tests in llvm that should allow you to use the c registers. And then `.arch morello` to turn it off again (or maybe `.arch morello+a64c`)

File: sys/arm64/arm64/support.S [hunk 5]
```diff
@@ -94,130 +126,167 @@ ENTRY(casueword)
...
+ */
+ENTRY(fuecap)
+	check_user_address 0, (VM_MAXUSER_ADDRESS-15), fsu_fault_nopcb
+	adr	x6, fsu_noswitch_fault	/* Load the fault handler */
+	SET_FAULT_HANDLER(x6, x2)	/* And set it */
+	SWITCH_MODE
+	ldtr	c0, [x0]		/* Try loading the data */
+	SWITCH_MODE
+	SET_FAULT_HANDLER(xzr, x2)	/* Reset the fault handler */
+	str	c0, [c1]		/* Save the data in kernel space */
```
Comment (jrtc27) [inline sys/arm64/arm64/support.S:203]: As Brett just discovered and fixed on morello-dev:
```suggestion
	str	c0, [x1]		/* Save the data in kernel space */
```

**Comment by zxombie** [comment]
I've pushed a few improvements upstream & am pulling them back as the first 4 commits in the new batch. This should make the `support.S` changes cleaner.

File: sys/arm64/arm64/copyinout.S [hunk 6]
```diff
@@ -91,28 +112,48 @@ END(copyin)
...
 	mov	w4, #1		/* If zero return faulure */
 	cbz	x2, 3f		/* If len == 0 then skip loop */
 
-	adr	x6, copyio_fault /* Get the handler address */
+#if __has_feature(capabilities)
+	cvtd	c1, x1			/* Store through a capability */
+	scbnds	c1, c1, x2		/* Set the bounds */
+#endif
+
+	adr	x6, copyio_c64_fault /* Get the handler address */
```
Comment (jrtc27) [inline sys/arm64/arm64/copyinout.S:128]: Hm, I don't like that we call it C64 even in non-CHERI kernels. But also, shouldn't it "just work" by keying off the low bit of the address in pcb_onfault (which will need a copy of the code used for signal handlers)?

Comment (jrtc27) [reply sys/arm64/arm64/copyinout.S:128]: Also it's a little dangerous to do this _before_ entering C64... you'd hope to not get a fault before then, but things will go off the rails if you do

File: sys/arm64/arm64/copyinout.S [hunk 7]
```diff
@@ -222,3 +264,172 @@ ending:
...
+ * lr - the return address, so jump here instead of calling
+ *
+ * In copytagcommon the bottom 4 bits of c0 and c1 must be identical so
+ * they can both be aligned load & store capabilities.
+ */
+	.text
+	.align	4
+	.local	copytagcommon
+	.type	copytagcommon,@function
+
```
Comment (jrtc27) [inline sys/arm64/arm64/copyinout.S:345]: copycapcommon please

File: sys/arm64/arm64/support.S [hunk 8]
```diff
@@ -145,8 +182,22 @@ EENTRY(fueword64)
...
+	str	c0, [x1]		/* Save the data in kernel space */
+	mov	x0, #0			/* Success */
+	ret				/* Return */
+END(fuecap)
+#endif
+
 /*
- * int subyte(volatile void *, int)
+ * int subyte(volatile void * __capability, int)
+ * TODO: __capability
```
Comment (jrtc27) [inline sys/arm64/arm64/support.S:200]: Are these really still TODO given the common macro in use?

File: sys/arm64/arm64/copyinout.S (hunk 6 above)
Comment (zxombie) [reply sys/arm64/arm64/copyinout.S:128]: There shouldn't be anywhere we could fault between setting the fault handler and entering C64 mode. The only instruction is the `ldr`, however as this is a pseudo instruction the assembler will use a `mov` so won't cause a trap.

It's possible PCC is set so the bounds are incorrect & we trap, but that should never happen, and if it does would indicate larger problems. Interrupts are also OK as they should work at any point in this code without changing it's behaviour.

Comment (jrtc27) [reply sys/arm64/arm64/copyinout.S:128]: But if you flip the order then it'll be safe and panic on an unexpected trap.

File: sys/arm64/arm64/support.S (hunk 8 above)
Comment (zxombie) [reply sys/arm64/arm64/support.S:200]: They are left over from when I was reworking the code & I missed them

File: sys/arm64/arm64/copyinout.S (hunk 6 above)
Comment (zxombie) [reply sys/arm64/arm64/copyinout.S:128]: If I flip the order the `SET_FAULT_HANDLER` macro won't work. The current order is safe as we need there to be specific types of aborts for the fault handler to be called that the kernel is careful to not cause.

Comment (jrtc27) [reply sys/arm64/arm64/copyinout.S:128]: It's just loads and stores, shouldn't the alternate mode ones work in C64?

Comment (zxombie) [reply sys/arm64/arm64/copyinout.S:128]: There is no alt mode `ldtrb`

Comment (zxombie) [reply sys/arm64/arm64/copyinout.S:128]: Sorry, mis-read what you wrote.

**Review by jrtc27 — CHANGES_REQUESTED** [review]
This currently has various bugs. As well as cleaning it up in general on morello-dev per my feedback, I also fixed the bugs in 16cd4176006b3a6f8d0fc2c39590fdcbfae7618b; see the commit for details.

**Comment by zxombie** [comment]
I don't understand the `copycommon` part of 16cd4176006b3a6f8d0fc2c39590fdcbfae7618b. `copy_main` should work when either the source or destination are unaligned as it is loading/storing to/from general purpose registers. We can perform these operations on unaligned memory, although it may be less efficient than if one or both are aligned.

**Comment by jrtc27** [comment]
> I don't understand the `copycommon` part of [16cd417](https://github.com/CTSRD-CHERI/cheribsd/commit/16cd4176006b3a6f8d0fc2c39590fdcbfae7618b). `copy_main` should work when either the source or destination are unaligned as it is loading/storing to/from general purpose registers. We can perform these operations on unaligned memory, although it may be less efficient than if one or both are aligned.

I was going off the fact that the code was:
https://github.com/CTSRD-CHERI/cheribsd/blob/7a2f2fa1381b93865b0f2339a33074da5af31a1c/sys/arm64/arm64/copyinout.S#L408-L413
which is clearly bogus (branching and falling through to the same place), has an inaccurate comment (arguments are free to be as unaligned with each other as you want) and implies that the _intent_ was to do a byte copy loop before the word copy loop, which is what I then implemented, though I probably did forget that AArch64 allows unaligned accesses like this (although it can still be configured to trap).

Also, the non-CHERI copycommon does a byte-by-byte copy if _either_ input is unaligned, rather than just blindly doing unaligned copies as you suggest is the right thing to do, so this should be more efficient than the upstream code.

Finally, yes, maybe the best thing to do in the case where everything is unaligned no matter what is to do block copies. But if given pointers to 0x1004 and 0x2004 you likely don't want _every_ iteration to do an unaligned load+store.

**Comment by zxombie** [comment]
There are a few issues in `external_abort`, a new abort type handler, however this should be in approximately the correct shape.

**Comment by jrtc27** [comment]
I see you've removed my byte copy loop prologue from copycommon. I'm still not convinced that's a good idea, but you should at least not attribute that to me, and update the commit message as it's currently wrong.

**Comment by zxombie** [comment]
The bogus code was copied from the capability copy version & was missed.

I would expect on the model it's best to prefer shorter code here to reduce the number of instructions it needs to emulate. On HW we would need to profile to find the best version of this code, e.g. it may be better to just perform a single unaligned copy, then use a partially aligned copy of the main data, and an unaligned copy of the trailing data. This would mean there is a double copy of some of the start and end of the partially aligned data, but this should be in cache, so should be fast.

**Comment by zxombie** [comment]
I've rebased on dev, updated the `Fixes to Morello copyinout` message, and added a build fix for `external_abort`. I expect @brooksdavis will pull in the upstream changes before me as they will be in the next week to be merged so am planning to rebase once more.

File: sys/arm64/include/cheric.h [hunk 9]
```diff
@@ -1 +1,62 @@
...
+		/*
+		 * TODO: add and initialize a distinct sealcap, remove
+		 * sealing perms from userspace_cap.
+		 */
+		elr = cheri_unseal(elr, cheri_setaddress(userspace_cap,
+		    CHERI_OTYPE_SENTRY));
+	}
+	if (elr & 0x1) {
+		tf->tf_spsr |= PSR_C64;
+		elr = cheri_incoffset(elr, -1);
```
Comment (arichardson) [inline sys/arm64/include/cheric.h:54]:
```suggestion
		elr--;
```
or (possibly less efficient)
```suggestion
		elr &= ~(uintcap_t)0x1;
```

Comment (zxombie) [reply sys/arm64/include/cheric.h:54]: @jrtc27 changed this to `--elr;` in a later commit. I haven't pulled it into this pull request as it's not needed by `copyin`, or it's variants.

**Comment by jrtc27** [comment]
As I’ve said before each commit should be self-contained and bug-free, with the possible exception of leaving holes for future commits to fill in if those are logically part of a different big set of changes. That means no fixup commits and no leaving out stylistic fixes. The whole point of this process is to get a close-to-pristine history that’s free from all the warts of its development, otherwise we might as well just give up and merge the tangled mess that is morello-dev directly.

File: sys/arm64/arm64/trap.c [hunk 10]
```diff
@@ -212,7 +225,6 @@ align_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
 	userret(td, frame);
 }
 
-
```
Comment (jrtc27) [inline sys/arm64/arm64/trap.c:215]: Unnecessary diff

**Comment by jrtc27** [comment]
The history still needs cleaning up; trapframe_set_elr is not present initially and so the history contains the crummy inlined copy that is later removed.

**Comment by zxombie** [comment]
The inlined version is not wrong. Reordering the commits will be a pain as the change to add trapframe_set_elr depends on the change to check pcb_onfault. The only issue I can see with the pcb_onfault change is it adds a nested check for capability support.

**Comment by jrtc27** [comment]
> The inlined version is not wrong. Reordering the commits will be a pain as the change to add trapframe_set_elr depends on the change to check pcb_onfault. The only issue I can see with the pcb_onfault change is it adds a nested check for capability support.

It's not functionally wrong, but this is landing a clean patch series into dev and so should get things right from the start rather than adding code that we know will later be moved. We should be making full use of our hindsight.

**Comment by jrtc27** [comment]
Full diff of the updated patch series:
```diff
diff --git a/sys/arm64/arm64/trap.c b/sys/arm64/arm64/trap.c
index 392baefc28ed..9e7243989f2f 100644
--- a/sys/arm64/arm64/trap.c
+++ b/sys/arm64/arm64/trap.c
@@ -226,6 +226,7 @@ align_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
 	userret(td, frame);
 }
 
+
 static void
 external_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
     uint64_t far, int lower)
@@ -237,7 +238,9 @@ external_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
 	 */
 	if (!lower && test_bs_fault((uintcap_t)frame->tf_elr)) {
 #if __has_feature(capabilities)
-		trapframe_set_elr(frame, (uintcap_t)generic_bs_fault);
+		trapframe_set_elr(frame,
+		    (uintcap_t)cheri_setaddress(cheri_getpcc(),
+		    (uint64_t)generic_bs_fault));
 #else
 		frame->tf_elr = (uint64_t)generic_bs_fault;
 #endif
@@ -254,7 +257,6 @@ static void
 cap_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
     uint64_t far, int lower)
 {
-	uintcap_t onfault;
 	struct pcb *pcb;
 
 	pcb = td->td_pcb;
@@ -262,14 +264,9 @@ cap_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
 		if (td->td_intr_nesting_level == 0 &&
 		    pcb->pcb_onfault != 0) {
 			frame->tf_x[0] = EPROT;
-			onfault = pcb->pcb_onfault;
-#if __has_feature(capabilities)
 			trapframe_set_elr(frame,
 			    (uintcap_t)cheri_setaddress(cheri_getpcc(),
-			    onfault));
-#else
-			frame->tf_elr = onfault;
-#endif
+			    pcb->pcb_onfault));
 			return;
 		}
 		print_registers(frame);
@@ -293,7 +290,6 @@ data_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
 	struct pcb *pcb;
 	vm_prot_t ftype;
 	int error, sig, ucode;
-	uintcap_t onfault;
 #ifdef KDB
 	bool handled;
 #endif
@@ -372,13 +368,12 @@ data_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
 			if (td->td_intr_nesting_level == 0 &&
 			    pcb->pcb_onfault != 0) {
 				frame->tf_x[0] = error;
-				onfault = pcb->pcb_onfault;
 #if __has_feature(capabilities)
 				trapframe_set_elr(frame,
 				    (uintcap_t)cheri_setaddress(cheri_getpcc(),
-				    onfault));
+				    pcb->pcb_onfault));
 #else
-				frame->tf_elr = onfault;
+				frame->tf_elr = pcb->pcb_onfault;
 #endif
 				return;
 			}
diff --git a/sys/arm64/include/cherireg.h b/sys/arm64/include/cherireg.h
index 37be7ad63746..fb191f8b52c7 100644
--- a/sys/arm64/include/cherireg.h
+++ b/sys/arm64/include/cherireg.h
@@ -147,4 +147,8 @@
 #define	CHERI_OTYPE_ISKERN(x)	(((x) & CHERI_OTYPE_KERN_FLAG) != 0)
 #define	CHERI_OTYPE_ISUSER(x)	(!(CHERI_OTYPE_ISKERN(x)))
 
+/* Reserved CHERI object types: */
+#define	CHERI_OTYPE_UNSEALED	(0l)
+#define	CHERI_OTYPE_SENTRY	(1l)
+
 #endif /* _ARM64_INCLUDE_CHERIREG_H_ */
diff --git a/sys/cheri/cherireg.h b/sys/cheri/cherireg.h
index 6f5377a54808..3e08bef156d0 100644
--- a/sys/cheri/cherireg.h
+++ b/sys/cheri/cherireg.h
@@ -115,10 +115,6 @@
     (CHERI_OTYPE_KERN_MAX - CHERI_OTYPE_KERN_MIN + 1)
 #define	CHERI_SEALCAP_KERNEL_OFFSET	0x0
 
-/* Reserved CHERI object types: */
-#define	CHERI_OTYPE_UNSEALED	(-1l)
-#define	CHERI_OTYPE_SENTRY	(-2l)
-
 #define	CHERI_REPRESENTABLE_LENGTH(len) \
 	__builtin_cheri_round_representable_length(len)
 #define	CHERI_REPRESENTABLE_ALIGNMENT_MASK(len) \
diff --git a/sys/mips/include/cherireg.h b/sys/mips/include/cherireg.h
index 30439393321f..f33ed5e746f1 100644
--- a/sys/mips/include/cherireg.h
+++ b/sys/mips/include/cherireg.h
@@ -153,6 +153,10 @@
 #define	CHERI_OTYPE_ISKERN(x)	(((x) & CHERI_OTYPE_KERN_FLAG) != 0)
 #define	CHERI_OTYPE_ISUSER(x)	(!(CHERI_OTYPE_ISKERN(x)))
 
+/* Reserved CHERI object types: */
+#define	CHERI_OTYPE_UNSEALED	(-1l)
+#define	CHERI_OTYPE_SENTRY	(-2l)
+
 /*
  * When performing a userspace-to-userspace CCall, capability flow-control
  * checks normally prevent local capabilities from being delegated.  This can
diff --git a/sys/riscv/include/cherireg.h b/sys/riscv/include/cherireg.h
index 4dfcb23d4e14..4bcfbfc83626 100644
--- a/sys/riscv/include/cherireg.h
+++ b/sys/riscv/include/cherireg.h
@@ -151,6 +151,10 @@
 #define	CHERI_OTYPE_ISKERN(x)	(((x) & CHERI_OTYPE_KERN_FLAG) != 0)
 #define	CHERI_OTYPE_ISUSER(x)	(!(CHERI_OTYPE_ISKERN(x)))
 
+/* Reserved CHERI object types: */
+#define	CHERI_OTYPE_UNSEALED	(-1l)
+#define	CHERI_OTYPE_SENTRY	(-2l)
+
 /*
  * List of CHERI capability cause code constants, which are used to
  * characterise various CP2 exceptions.
```
This folds in some cleanups made on morello-dev (I think I did them at least?), includes the sentry commits from #842 as they're needed here for trapframe_set_elr, adds additional cleanups and bug fixes (external_abort would have broken on ERET if ever used due to not deriving a capability).

Please check the commit messages still make sense after the reordering (I tweaked the trapframe_set_elr one a bit as it's now added at the same time as the first correct bit of tf_elr handling).

NB: The whitespace change is intentional; the patch series here removed that extraneous newline and I have removed the removal to reduce the diff to upstream.

**Comment by jrtc27** [comment]
We currently have this diff to morello-dev in copyinout.S (ignoring a whitespace change and the decision to not have a byte copy prologue for the plain copycommon):
```diff
diff --git a/sys/arm64/arm64/copyinout.S b/sys/arm64/arm64/copyinout.S
index af0605bc68b6..cdb0346b2b7d 100644
--- a/sys/arm64/arm64/copyinout.S
+++ b/sys/arm64/arm64/copyinout.S
@@ -52,29 +52,23 @@ __FBSDID("$FreeBSD$");
     ldr x7, =VM_MAXUSER_ADDRESS
     cmp x6, x7
     b.hi    \bad_access_func
 .endm
 
 /*
  * Fault handler for the copy{in,out} functions below.
  */
-#if __has_feature(capabilities)
-    .arch_extension c64
-#endif
-ENTRY(copyio_c64_fault)
-    EXIT_C64
-EENTRY(copyio_fault)
+ENTRY(copyio_fault)
     SET_FAULT_HANDLER(xzr, x1) /* Clear the handler */
     EXIT_USER_ACCESS_CHECK(w0, x1)
 copyio_fault_nopcb:
     mov x0, #EFAULT
     ret
-EEND(copyio_fault)
-END(copyio_c64_fault)
+END(copyio_fault)
 
 /*
  * Copies from a kernel to user address
  *
  * int copyout(const void *kaddr, void * __capability udaddr, size_t len)
  */
 ENTRY(copyout)
     cbz x2, 1f
@@ -135,17 +129,17 @@ ENTRY(copyinstr)
     ENTER_C64
 1:
 #if __has_feature(capabilities)
     gcvalue x9, c0
     cmp x9, x7
 #else
     cmp x0, x7
 #endif
-    b.cs    copyio_c64_fault
+    b.cs    copyio_fault
 #if __has_feature(capabilities)
     ldtrb   w4, [c0]    /* Load from uaddr */
     add c0, c0, #1  /* Next char */
     strb    w4, [c1], #1    /* Store in kaddr */
 #else
     ldtrb   w4, [x0]    /* Load from uaddr */
     add x0, x0, #1  /* Next char */
     strb    w4, [x1], #1    /* Store in kaddr */
```
That looks important to me? I remember wanting to get rid of copyio_c64_fault when rewriting copyinout back in October but being unable to due to copyinstr. @zxombie?

File: sys/arm64/arm64/support.S [hunk 11]
```diff
@@ -38,16 +38,26 @@ __FBSDID("$FreeBSD$");
...
 	cmp	x\user_arg, x7
+#endif
 	b.cs	\bad_addr_func
 .endm
 
 /*
  * One of the fu* or su* functions failed, return -1.
  */
 ENTRY(fsu_fault)
+	EXIT_C64
```
Comment (zxombie) [inline sys/arm64/arm64/support.S:60]: This shouldn't be needed any more.

**Comment by zxombie** [comment]
The copyio diff to morello-dev is because we have the old version in that branch. It will be replaced with this on the next merge.

File: sys/arm64/arm64/support.S (hunk 11 above)
Comment (jrtc27) [reply sys/arm64/arm64/support.S:60]: Agreed, that got dropped on morello-dev but not here.

**Comment by jrtc27** [comment]
> The copyio diff to morello-dev is because we have the old version in that branch. It will be replaced with this on the next merge.

I don't understand how this one is supposed to work though? This is a direct branch executing in C64 to a function that's A64, we need to manually leave A64 because it's neither a `BX #4` nor an indirect capability branch that will automatically set PSTATE.C64, otherwise we'll execute the A64 code as C64.

File: sys/arm64/arm64/trap.c [hunk 12]
```diff
@@ -168,16 +168,17 @@ extern uint32_t generic_bs_poke_1f, generic_bs_poke_2f;
...
 {
-	return (addr == &generic_bs_peek_1f ||
-	    addr == &generic_bs_peek_2f ||
-	    addr == &generic_bs_peek_4f ||
-	    addr == &generic_bs_peek_8f ||
-	    addr == &generic_bs_poke_1f ||
-	    addr == &generic_bs_poke_2f ||
-	    addr == &generic_bs_poke_4f ||
-	    addr == &generic_bs_poke_8f);
+	/* TODO: Will need fixing for a purecap kernel */
```
Comment (bsdjhb) [inline sys/arm64/arm64/trap.c:186]: How will this need fixing for purecap?  It seems like it will work fine for purecap?

File: sys/arm64/arm64/trap.c [hunk 13]
```diff
@@ -239,6 +252,35 @@ external_abort(struct thread *td, struct trapframe *frame, uint64_t esr,
...
+			    pcb->pcb_onfault));
+			return;
+		}
+		print_registers(frame);
+		printf(" far: %16lx\n", far);
+		printf(" esr:         %.8lx\n", esr);
+		panic("Capability abort from kernel space!");
+	}
+
+	call_trapsignal(td, SIGPROT, 0, (void * __capability)frame->tf_elr,
```
Comment (bsdjhb) [inline sys/arm64/arm64/trap.c:278]: I guess a later change pulls in the `si_code` for `SIGPROT`?  I would kind of expect that helper function for `si_code` to be included here and for `si_code` to be correct from the start.

File: sys/arm64/arm64/trap.c (hunk 12 above)
Comment (jrtc27) [reply sys/arm64/arm64/trap.c:186]: Function pointer address will have the LSB set, but ELR will not even if trapping from C64, that info goes in SPSR.C64 only.

File: sys/arm64/arm64/trap.c (hunk 13 above)
Comment (jrtc27) [reply sys/arm64/arm64/trap.c:278]: Hm, yes, that commit belongs here not in #842. There's be a slight annoying rebase of #842 needed once this PR is merged with that in order to resolve the conflicts in files.arm64 and cheri.h but that's fine.

Comment (bsdjhb) [reply sys/arm64/arm64/trap.c:278]: Alternatively, you could just move this commit to the end of #842 and patch it to use the function to map the `si_code`.

File: sys/arm64/include/asm.h [hunk 14]
```diff
@@ -60,6 +60,22 @@
 /* Alias for link register x30 */
 #define	lr		x30
 
+/*
+ * Switch into C64 mode to use instructions only available in Morello.
+ */
+#if __has_feature(capabilities) && !defined __CHERI_PURE_CAPABILITY__
```
Comment (jrtc27) [inline sys/arm64/include/asm.h:66]:
```suggestion
#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
```

File: sys/arm64/arm64/trap.c (hunk 13 above)
Comment (jrtc27) [reply sys/arm64/arm64/trap.c:278]: Yeah, perhaps having copyinout in a separate PR just doesn't make sense given it's all related.

**Review by brooksdavis — APPROVED** [review]

**Comment by jrtc27** [comment]
(this will be folded into #842 as commits need to move between them, much of this is only incidentally relevant to copyinout and it really doesn't make that much sense to try and keep things separate)
